src/Components/Web.JS/src/FileDownloader.ts

```diff
@@ -0,0 +1,14 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+export async function downloadFile(data: any, fileName: string) {
+    const arrayBuffer = await data.arrayBuffer();
+    const blob = new Blob([arrayBuffer]);
+    const url = URL.createObjectURL(blob);
+    const anchorElement = document.createElement('a');
+    anchorElement.href = url;
+    anchorElement.download = fileName;
+    anchorElement.click();
+    anchorElement.remove();
+    URL.revokeObjectURL(url);
+}
```

Comment:
It's great that you've implemented `downloadFile` using `Blob` and `createObjectURL` since that's the only way that works on all browsers. We do need this.

However, this approach does have the drawback that it involves:

[1] Buffering the entire file in browser memory, even if it's gigabytes
[2] Not showing the "save as" dialog until we've got the while file data into browser memory

On Chromium browsers (and probably other browsers in the future), there's a new spec called the FileSystem API that provides a better option, whereby we can show the "save as" dialog first, and then stream the data into the file without buffering it all in memory. It works like this:

```js
// This is actually supplied by JS interop - no need to construct a ReadableStream in the new code
const readableStream = new Response(new TextEncoder().encode('Some contents here')).body;

// Show the "save as dialog"
let fileWriter;
try {
    const fileHandle = await window.showSaveFilePicker();
    fileWriter = await fileHandle.createWritable();
} catch {
    // User pressed cancel, so abort the whole thing
    return;
}

const reader = readableStream.getReader();
while (true) {
    const readResult = await reader.read();
    if (readResult.done) {
        break;
    }

    await fileWriter.write(readResult.value);
}

await fileWriter.close();
```

You can detect whether or not the FileSystem API is available in the current browser by checking if `typeof(window.showSaveFilePicker) === 'function'`.

So, in order to provide a better experience on Chromium (and future) browsers, do you think we could detect and `window.showSaveFilePicker` and switch between the two implementations based on that?

Reply:
BTW I'm aware this means the user won't see the file showing up in their "Downloads" UI within the browser. But if the Blob approach doesn't really solve that either because the download doesn't happen as far as the browser is concerned until the JS code has actually downloaded all the data and buffered it in memory. So I still think the UX is better with the FileSystem API implementation overall.

Reply:
Agree, this would be nice to have.

Do we think we need some sort of way to have progress reporting? I can think that we could do this directly on the .NET side. (Technically users could do this themselves) by wrapping the Stream and firing up an event as it is being read. So, a component could register for that event and update the UI accordingly (Imagine a toast/progress bar type of thing).

That would to a degree for both cases (new API + old API). That said, if we can optionally leverage the new API, that would be great (and orthogonal to my suggestion)

Reply:
Maybe this is the basis for having a component for this in the future, imagine something like `<DownloadableFile Source="..." />` that displays a button to start the download, and switches to render a progress bar in place as the download is happening. (not saying we have to do it, just throwing the idea out there)

Reply:
> Do we think we need some sort of way to have progress reporting? ... Technically users could do this themselves

The fact that people can already do this themselves on the .NET side makes it fine to not bake in any secondary feature around progress status right now. Let's not expand the scope of the feature too much.

> imagine something like <DownloadableFile Source="..." /> that displays a button to start the download, and switches to render a progress bar in place as the download is happening ... not saying we have to do it

Agreed that would be cool, but also agreed that it doesn't have to be baked-in as a framework feature when people can do it themselves. Maybe in the long term if there's clear enough customer demand though!

Reply:
Thanks Steve, wasn't aware of that API. Took a brief look and another couple of things to note for `showSaveFilePicker`:

- Chromium **desktop** exclusive (doesn't work on mobile)
- Requires secure context

[`caniuse`](https://caniuse.com/?search=showSaveFilePicker) has this at 27% of mobile users and 75% of desktop users. Anecdotally, are more likely to be downloading on a Desktop than a mobile device, I still think this would be worthwhile.